llm/training/config.py

```diff
@@ -0,0 +1,341 @@
+DatasetConfig.register("load_dataset", load_dataset)
+
+
+def default_training_arguments(
+    eval_dataset: Optional[DatasetConfig],
+    experiment_tracking_config: Optional[ExperimentTrackingConfig] = None,
+) -> Dict[str, Any]:
+    defaults: Dict[str, Any] = dict(
+        bf16=False,
+        optim="adamw_torch",
+        num_train_epochs=1,
+        gradient_accumulation_steps=2,
+        per_device_train_batch_size=2,
+        per_device_eval_batch_size=2,
+        auto_find_batch_size=False,
+        gradient_checkpointing=True,
+        logging_strategy="steps",
+        logging_steps=10,
+        save_steps=30,
+        save_strategy="steps",
+    )
+    if eval_dataset:
+        defaults.update(dict(eval_steps=5, evaluation_strategy="steps"))
+    if experiment_tracking_config:
+        defaults.update({"report_to": experiment_tracking_config.report_to})
+    return defaults
+
+
+@dataclass
+class FineTuneConfig:
+    base_model: str
+    train_dataset_config: DatasetConfig
+    training_arguments: TrainingArguments
+    lora_config: LoraConfig
+    experiment_tracking_config: Optional[ExperimentTrackingConfig] = None
+    eval_dataset_config: Optional[DatasetConfig] = None
+
+    # we need a local directory on the machine to save files.
+    local_output: Optional[str] = None
+    # we can additionally copy data to other networked locations.
+    additional_output_paths: List[str] = field(default_factory=list)
+
+    torch_dtype: torch.dtype = torch.float16
+    load_in_4bit: bool = False
+    load_in_8bit: bool = False
+    use_gradient_checkpointing: bool = True
+    resume_from_checkpoint: bool = True
+
+    @classmethod
+    def from_config(cls, **config) -> "FineTuneConfig":
+        model_config = ModelConfig.from_registry(config["base_model"])
+        default_lora_dict = model_config.default_lora_config.copy()
+
+        if config.get("local_output") is None:
+            f = tempfile.mkdtemp()
+            config["local_output"] = f
+
+        experiment_tracking_config = load_config(
+            ExperimentTrackingConfig, config.pop("experiment_tracking_config", None)
+        )
+        train_dataset_config = load_config(DatasetConfig, config.pop("train_dataset_config"))
+        eval_dataset_config = load_config(DatasetConfig, config.pop("eval_dataset_config"))
+
+        if "torch_dtype" in config:
+            config["torch_dtype"] = getattr(torch, config["torch_dtype"])
+
+        training_arguments_dict = default_training_arguments(
+            eval_dataset_config, experiment_tracking_config
+        )
+        training_arguments_dict.update(config.pop("training_arguments", {}))
+        training_arguments_dict["output_dir"] = config["local_output"]
+        training_arguments = load_config(TrainingArguments, training_arguments_dict)
+
+        lora_config_dict = default_lora_dict
+        lora_config_dict.update(config.pop("lora_config", {}))
+        lora_config = load_config(LoraConfig, lora_config_dict)
```

Comment:
Should there be a way to load fine tuning config without peft/lora?

Reply:
hm. yes, but not yet - I'm not sure how much we have to change this script to fine tune well without lora